src/daemons/linuxd/src/linuxd/mod.rs

```diff
@@ -814,105 +873,60 @@ impl LinuxDaemon {
-                        debug!("read {count} bytes from the gateway");
-
-                        // Truncate read request to fit in the response buffer.
-                        let read_count: usize = if count > request.count as usize {
-                            warn!(
-                                "handle_read_request(): truncating payload (requested={}, \
-                                 actual={count})",
-                                { request.count },
-                            );
-                            request.count as usize
-                        } else {
-                            count
-                        };
-
-                        let mut response_buf: [u8; ReadResponse::BUFFER_SIZE] =
-                            [0u8; ReadResponse::BUFFER_SIZE];
-                        response_buf[..read_count].copy_from_slice(&message[..read_count]);
-                        profiler::timestamp_message!(&mut response_buf, 0);
-
-                        // Check if there are any outstanding bytes to be read.
-                        if count > read_count {
-                            // Break outstanding bytes into multiple read responses.
-                            for i in (read_count..count).step_by(ReadResponse::BUFFER_SIZE) {
-                                let end: usize = i + ReadResponse::BUFFER_SIZE;
-                                let end: usize = if end > count { count } else { end };
-                                let mut response_buf: [u8; ReadResponse::BUFFER_SIZE] =
-                                    [0u8; ReadResponse::BUFFER_SIZE];
-                                response_buf[..end - i].copy_from_slice(&message[i..end]);
-                                env.push_stdin_message(ReadResponse::build(
-                                    source,
-                                    (end - i) as c_ssize_t,
-                                    response_buf,
-                                ));
-                            }
-                        }
-                        // Push EoF message.
-                        // When timestamping messages as part of data-path profiling, we require
-                        // applications to not expect an EoF when reading from stdin. This is
-                        // because the profiling modifies the payload of all messages, including
-                        // EoFs.
-                        #[cfg(not(feature = "timestamp-messages"))]
-                        env.push_stdin_message(ReadResponse::build(
-                            source,
-                            0,
-                            [0u8; ReadResponse::BUFFER_SIZE],
-                        ));
-
-                        ReadResponse::build(source, read_count as c_ssize_t, response_buf)
+            // Send ReadRequest to gateway IO thread.
+            if let Err(error) = gateway_stdin_tx.send((request, env.get_stdin_response_tx())) {
```

Comment:
We `send()` while holding the lock. If `send()` is blocked, do we have any unwanted side effects?

Reply:
Good catch. Seems unnecesarily risky. I have moved the send to after the match statement, once we have release the lock.